File: frontend/src/api/customUrl.ts
import { apiClient } from "./client";

export interface CustomUrlTask {
  id: number;
  url: string;
  domain?: string;
  platform?: string;
  title?: string;
  description?: string;
  content_excerpt?: string;
  summary?: string;
  keywords: string[];
  sentiment?: string;
  category?: string;
  trend_opinion?: string;
  status: "pending" | "fetching" | "analyzing" | "completed" | "failed" | "blocked";
  error_message?: string;
  created_at: string;
  updated_at: string;
}

export interface CustomUrlAnalyzeResponse {
  task_id: number;
  status: CustomUrlTask["status"];
  title?: string;
  platform?: string;
  summary?: string;
  keywords: string[];
  sentiment?: string;
  category?: string;
  trend_opinion?: string;
  source_url: string;
}

export function submitCustomUrl(url: string) {
  return apiClient.post<CustomUrlTask>("/custom-url/submit", { url });
}

export function fetchCustomUrlTasks() {
  return apiClient.get<CustomUrlTask[]>("/custom-url/tasks");
}

export function fetchCustomUrlTask(taskId: number) {
  return apiClient.get<CustomUrlTask>(`/custom-url/tasks/${taskId}`);
}

export function analyzeCustomUrl(url: string) {
  return apiClient.post<CustomUrlAnalyzeResponse>("/custom-url/analyze", { url });
}

File: backend/app/schemas/custom_url.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class CustomUrlSubmit(BaseModel):
    url: HttpUrl


class CustomUrlAnalyzeRequest(BaseModel):
    url: HttpUrl


class CustomUrlTaskRead(BaseModel):
    id: int
    url: str
    domain: str | None = None
    platform: str | None = None
    title: str | None = None
    description: str | None = None
    content_excerpt: str | None = None
    summary: str | None = None
    keywords: list[str] = Field(default_factory=list)
    sentiment: str | None = None
    category: str | None = None
    trend_opinion: str | None = None
    status: str
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CustomUrlAnalyzeResponse(BaseModel):
    task_id: int
    status: str
    title: str | None = None
    platform: str | None = None
    summary: str | None = None
    keywords: list[str] = Field(default_factory=list)
    sentiment: str | None = None
    category: str | None = None
    trend_opinion: str | None = None
    source_url: str
